### ISSUE TITLE
Template.Tests consistently fail on Windows due to selenium and npm failures

### ISSUE THREAD

**Issue by natemcmaster** [aspnet/templating#131]
I have not been able to make the tests pass on Windows. The tests are frequently failing with one of two issues. Any ideas what to change?

1. Several tests fail when Seleium fails to start

```
Templates.Test.SpaTemplateTest.SpaTemplate_Works_NetCore(template: "vue") [FAIL]
OpenQA.Selenium.WebDriverException : Cannot start the driver service on http://localhost:64591/
Stack Trace:
            at OpenQA.Selenium.DriverService.Start()
            at OpenQA.Selenium.Remote.DriverServiceCommandExecutor.Execute(Command commandToExecute)
            at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
            at OpenQA.Selenium.Remote.RemoteWebDriver.StartSession(ICapabilities desiredCapabilities)
            at OpenQA.Selenium.Remote.RemoteWebDriver..ctor(ICommandExecutor commandExecutor, ICapabilities desiredCapabilities)        C:\dev\aspnet\Universe2\modules\Templating\test\Templates.Test\Helpers\WebDriverFactory.cs(23,0): at Templates.Test.Helpers.WebDriverFactory.CreateWebDriver()
        C:\dev\aspnet\Universe2\modules\Templating\test\Templates.Test\Helpers\AspNetProcess.cs(103,0): at Templates.Test.Helpers.AspNetProcess.VisitInBrowser()
```

2. Perpetual npm issues.

```
Failed   Templates.Test.SpaTemplateTest.SpaTemplate_Works_NetCore(template: "aurelia")

....

 ==> cmd /c "npm install" [C:\dev\aspnet\Universe2\modules\Templating\test\Templates.Test\bin\Debug\netcoreapp2.0\TestTemplates\bd302f03473b47c79813574f3739e840]
 [ERROR] npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but npm-shrinkwrap.json was generated for lockfileVersion@0. I'll try to do my best with it!
 [ERROR] npm WARN deprecated coa@1.0.3: Please upgrade to 1.0.4 for node 0.10, 0.12, or to 2.0+ for node 4+
 [ERROR] npm ERR! path C:\dev\aspnet\Universe2\modules\Templating\test\Templates.Test\bin\Debug\netcoreapp2.0\TestTemplates\bd302f03473b47c79813574f3739e840\node_modules\fsevents\node_modules\dashdash\node_modules
 [ERROR] npm ERR! code EPERM
 [ERROR] npm ERR! errno -4048
 [ERROR] npm ERR! syscall scandir
 [ERROR] npm ERR! Error: EPERM: operation not permitted, scandir 'C:\dev\aspnet\Universe2\modules\Templating\test\Templates.Test\bin\Debug\netcoreapp2.0\TestTemplates\bd302f03473b47c79813574f3739e840\node_modules\fsevents\node_modules\dashdash\node_modules'
 [ERROR] npm ERR!  { Error: EPERM: operation not permitted, scandir 'C:\dev\aspnet\Universe2\modules\Templating\test\Templates.Test\bin\Debug\netcoreapp2.0\TestTemplates\bd302f03473b47c79813574f3739e840\node_modules\fsevents\node_modules\dashdash\node_modules'
 [ERROR] npm ERR!   stack: 'Error: EPERM: operation not permitted, scandir \'C:\\dev\\aspnet\\Universe2\\modules\\Templating\\test\\Templates.Test\\bin\\Debug\\netcoreapp2.0\\TestTemplates\\bd302f03473b47c79813574f3739e840\\node_modules\\fsevents\\node_modules\\dashdash\\node_modules\'',
 [ERROR] npm ERR!   errno: -4048,
 [ERROR] npm ERR!   code: 'EPERM',
 [ERROR] npm ERR!   syscall: 'scandir',
 [ERROR] npm ERR!   path: 'C:\\dev\\aspnet\\Universe2\\modules\\Templating\\test\\Templates.Test\\bin\\Debug\\netcoreapp2.0\\TestTemplates\\bd302f03473b47c79813574f3739e840\\node_modules\\fsevents\\node_modules\\dashdash\\node_modules' }
 [ERROR] npm ERR!
 [ERROR] npm ERR! Please try running this command again as root/Administrator.
 [ERROR]
 [ERROR] npm ERR! A complete log of this run can be found in:
 [ERROR] npm ERR!     C:\Users\namc\AppData\Roaming\npm-cache\_logs\2017-10-26T23_36_51_972Z-debug.log
```

## My machine

dotnet: 2.0.3-servicing-007037
npm: 5.3.0
node: 8.1.3

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.16299
 OS Platform: Windows
 RID:         win10-x64

cc @jkotalik - who was also unable to get tests passing on his machine

**Comment by Eilon** [comment]
@SteveSandersonMS any idea on this? Any special instructions needed?

**Comment by SteveSandersonMS** [comment]
For issue 2, does "perpetual" mean "failing in the same way every time" or just "frequent but intermittent"? If it's intermittent, you probably need to ensure Windows Defender excludes the directory where NPM is creating and moving thousands of small files, otherwise if it's holding a lock on a file on the wrong moment, you do get those kinds of access denied errors. NPM has zero tolerance for temporary file access issues. If it's consistent, it might be path-length limitations, though I doubt it because the path in your error message is only 190 chars long.

For issue 1, could this be that you're on Windows 10 Creators Fall Update? As per our conversation before, the Edge driver for Selenium didn't appear to support that OS, so we need to either update our Selenium dependency if they have published a fix, or modify the logic about detecting browser automation support to return `false` if you're on Fall Update.

**Comment by natemcmaster** [comment]
Issue 2 - frequent but intermittent. It is frequently failing with "EPERM: operation not permitted, scandir", but the folder it shows in the error message changes. It's not Windows Defender: I have an exclusion for the folder containing the tests.

Side note: I wonder if this means `npm install` is also frequently failing for our users. In my hobby projects, I switched to yarn because NPM continually had these kinds of errors, even in npm5.

Issue 1 - that's most likely it. Selenium.WebDriver.MicrosoftDriver 16.16299.0 was released last week and appears to resolve it.

**Comment by SteveSandersonMS** [comment]
Looks like the `scandir` issue is worse than the usual NPM random glitches on Windows: https://github.com/npm/npm/issues/17747

You might want to downgrade to NPM 4.x until the 5.x line has a proper fix for this.

@natemcmaster If you agree that this is specifically an issue in NPM that we can't directly do anything about, I'd recommend we close this as an external issue.

**Comment by natemcmaster** [comment]
Yeah, it seems there isn't much we can do about npm's issues, except perhaps, abandon it for yarn. But let's consider that a separate issue.

You can leave this open by assign to me to ensure #136 get's ported to dev.

**Comment by SteveSandersonMS** [comment]
OK, assigning to you.

We can use `yarn` in our internal build infrastructure if we want; we just have to not rely on it as part of templates (as in, we shouldn't assume that people using the templates to create projects have it installed).

**Comment by natemcmaster** [comment]
Would adding this package to templates be reasonable? It would remove the need for a global install of yarn. https://github.com/natemcmaster/Yarn.MSBuild And if not, would it be okay to add it as a flag? `dotnet new react --with-yarn`?

**Comment by SteveSandersonMS** [comment]
Using `Yarn.MSBuild` would still force all developers to install their own copy of `yarn`, unless I'm mistaking how it works. `yarn` and `npm` aren't exactly compatible replacements for each other - for example, if you're using `yarn` then you're keeping lock information in `yarn.lock`, and if you're using `npm` then you're keeping it in `npm-shrinkwrap.json` (or the NPM 5 equivalent). Neither tool understands the other's lockfile format. So if we have MSBuild targets that invoke `yarn`, things will only behave consistently if the developer is also using `yarn` themselves when they do package installations/upgrades. Since this forces us to pick one of the two, it's safer to pick `npm` because everyone who has Node also has NPM.

> And if not, would it be okay to add it as a flag?

That would be possible, but then we have a new possible failure condition where the two lockfiles are inconsistent and our tests give false positives (e.g., we have the correct version for something in `yarn.lock` but the wrong version in `npm-shrinkwrap.json`).

If it's essential to use Yarn to have our tests work reliably, my first inclination would be something more like automating the process of generating a `yarn.lock` file from the `npm-shrinkwrap.json` during test execution, guaranteeing there's only one source of truth about versions.

**Comment by natemcmaster** [comment]
> Using Yarn.MSBuild would still force all developers to install their own copy of yarn, unless I'm mistaking how it works

Yarn is bundled in that package, no global install needed.

Yeah, I get there are bigger concerns with switching to yarn as the default. Probably not the right move at the moment. And for the sake of test reliability, it might be more work than it's worth. Thinking about this more, it's probably worth getting community feedback on this before we invest the effort. I'll open an separate issue and let's see if we get any kind of feedback on it.

**Comment by SteveSandersonMS** [comment]
> I'll open an separate issue and let's see if we get any kind of feedback on it.

I'd be cautious about reading much into what replies you get from posting an issue. Anyone posting is self-selected to be a highly interested and involved developer, exactly the kind who's likely to have already installed `yarn` or is willing to do so. The other 90% have never heard of `yarn` and won't reply to the issue.

**Comment by natemcmaster** [comment]
Yeah good points. You're the subject-matter expert here so I trust your judgment. I just wanted to suggest the idea independently of our test-reliability discussion on this thread.

**Comment by natemcmaster** [comment]
Fixed in #148
